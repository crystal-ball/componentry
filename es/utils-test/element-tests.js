/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { mount } from 'enzyme';
import dt from './dt';
/**
 * Library default component requirment test suite.
 *
 * IMPORTANT NOTES
 * - Enzyme `find()` will find **EVERY** instance of a prop, this makes it difficult
 *   to assert on find lengths, because it will be variable for components with
 *   HOCs which is why we use `toBeTruthy` to only check that some instance of the
 *   expected element attr exists
 */

export default (function (TestComponent) {
  var enzymeOptions = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  /*
   * Test that any html tag or component can be passed through the 'as' prop to
   * override the parent element the component renders
   */
  test('should render as specified html element or component', function () {
    // Create a component to validate that the TestComponent returns.
    var TestAs = function TestAs() {
      return /*#__PURE__*/React.createElement("section", null);
    };

    var wrapper = mount( /*#__PURE__*/React.createElement(TestComponent, {
      as: TestAs
    }), enzymeOptions);
    expect(wrapper.find(TestAs).length).toBeTruthy();
  });
  /*
   * All components should spread passed props on the parent container for max
   * customization
   */

  test('should pass through classnames and attributes', function () {
    var wrapper = mount( /*#__PURE__*/React.createElement(TestComponent, {
      className: "test-custom",
      "data-test": "test-custom"
    }), enzymeOptions);
    expect(wrapper.find('.test-custom').length).toBeTruthy();
    expect(wrapper.find(dt('test-custom')).length).toBeTruthy();
  });
});