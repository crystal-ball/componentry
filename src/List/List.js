import { Children } from 'react'
import elem from '../elem-factory'
import { useTheme } from '../Theme/Theme'

/**
 * [List component üìù](https://componentry.design/components/list)
 */
export default function List(props) {
  const { flush = false, ...rest } = { ...useTheme('List'), ...props }

  // Lists check first child (only first child!) to see if it has actions, and
  // if it does, it's not a `li` so we render a `ul`.
  const child = Children.toArray(props.children)[0]
  // We have to be careful of children like text nodes with this check
  const childProps = child && child.props ? child.props : {}

  return elem({
    as: childProps.href || childProps.onClick ? 'div' : 'ul',
    elemClassName: ['list', { 'list-flush': flush }],
    ...rest,
  })
}
List.displayName = 'List'

/**
 * [List item component üìù](https://componentry.design/components/list)
 */
List.Item = function ListItem(props) {
  const { active = false, color = '', ...rest } = { ...useTheme('ListItem'), ...props }
  const { href, onClick } = rest

  return elem({
    /* eslint-disable no-nested-ternary */
    as: href || onClick ? (href ? 'a' : 'button') : 'li',
    // List items always have the `list-item` class, and items that are
    // actionable (buttons/anchors) have the `list-action-item` modifier
    // class added with additional required styles.
    elemClassName: {
      active,
      'disabled': rest.disabled,
      'list-item': true,
      'list-action-item': href || onClick,
      [`list-item-${color}`]: color,
    },
    ...rest,
  })
}
List.Item.displayName = 'ListItem'
