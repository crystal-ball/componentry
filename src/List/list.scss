/* ~componentry/src/List/list */

// --------------------------------------------------------
// Lists (<ul>, <ol>, or <div>)

.list {
  display: flex;
  flex-direction: column;

  // No need to set list-style: none; since .list-item is block level
  padding-left: 0; // reset padding because ul and ol
  margin-bottom: 0;
}

// --------------------------------------------------------
// List items (<li> or <div> within .list)

.list-item {
  position: relative;
  display: block;
  padding: $list-item-padding-y $list-item-padding-x;
  background-color: $list-bg;
  border: $list-border-width solid $list-border-color;

  // For subsequent list item's remove the border top so we don't double up
  & + .list-item {
    border-top: 0;
  }

  &:first-child {
    @include border-top-radius($list-border-radius);
  }

  &:last-child {
    margin-bottom: 0;
    @include border-bottom-radius($list-border-radius);
  }

  &.disabled {
    color: $list-disabled-item-color;
    background-color: $list-disabled-item-bg;
  }
}

// --------------------------------------------------------
// Action list items
//
// Use anchor or button elements instead of `li`s or `div`s to create
// interactive list items. Includes an extra `.active` modifier class for
// selected items.

.list-action-item {
  color: $list-action-item-color;
  font-size: inherit; // -> buttons
  text-align: inherit; // -> buttons
  text-decoration: none; // -> anchors

  // Hover state
  &:hover,
  &:focus {
    z-index: 1; // Place hover/focus items above their siblings for proper border styling
    color: $list-action-item-hover-color;
    background-color: $list-action-item-hover-bg;
    text-decoration: none;
  }

  // Note that the active state is different from having the active class
  // (Class is persistent, state is for user interaction)
  &:active {
    color: $list-action-item-active-color;
    background-color: $list-action-item-active-bg;
  }
}

// --------------------------------------------------------
// Active state modifier

// Must be declared after list-action-item ensure active has precedence
// over :hover (Note this is valid for items and action items)
.list-item.active {
  z-index: 2; // Place active items above their siblings for proper border styling
  color: $list-active-item-color;
  background-color: $list-active-item-bg;
  border-color: $list-active-item-border-color;
}

// --------------------------------------------------------
// Flush layout list item

// Remove borders and border-radius to keep list group items edge-to-edge. Most
// useful within other components (e.g., cards).
.list-flush {
  .list-item {
    border-right: 0;
    border-left: 0;
    @include border-radius(0);
  }

  &:first-child {
    .list-item:first-child {
      border-top: 0;
    }
  }

  &:last-child {
    .list-item:last-child {
      border-bottom: 0;
    }
  }
}

// --------------------------------------------------------
// Contextual variants

// Add modifier classes to change text and background color on individual items.
// (Organizationally, this must come after the `:hover` states)
@each $color, $value in $theme-colors {
  $background: theme-color-level($color, -9);
  $foreground: theme-color-level($color, 6);

  .list-item-#{$color} {
    color: $color;
    background-color: $background;

    &.list-action-item {
      @include hover-focus {
        color: $foreground;
        background-color: darken($background, 5%);
      }

      &.active {
        color: theme-color(background);
        background-color: $foreground;
        border-color: $foreground;
      }
    }
  }
}
