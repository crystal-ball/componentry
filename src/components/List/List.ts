import { useTheme } from '../Theme/Theme'
import { ComponentBaseProps } from '../../utils/types'
import { element } from '../../utils/element-creator'

export interface ListProps extends ComponentBaseProps<'div'> {
  /** Removes list margins  */
  flush?: boolean
}

export interface ListItemProps extends ComponentBaseProps<'div'> {
  /** Sets active styles */
  active?: boolean
  /** Sets the item as disabled */
  disabled?: boolean
  /** Item href */
  href?: string
  /** Variant color */
  color?: 'primary'
  /** Item routing location */
  to?: string
}

export interface List {
  (props: ListProps): React.ReactElement
  displayName: 'List'
  /**
   * [Input description component üìù](https://componentry.design/components/input)
   */
  Item: React.FC<ListItemProps>
}

/**
 * [List component üìù](https://componentry.design/components/list)
 * @experimental
 */
export const List: List = (props) => {
  const { flush, ...rest } = { ...useTheme('List'), ...props }

  return element('List', {
    componentCx: { 'list-flush': flush },
    ...rest,
  })
}
List.displayName = 'List'

/**
 * [List item component üìù](https://componentry.design/components/list)
 */
List.Item = function ListItem(props) {
  const { active, color, disabled, ...rest } = { ...useTheme('ListItem'), ...props }

  return element('ListItem', {
    as: rest.href || rest.to ? 'a' : 'button',
    componentCx: {
      [`list-item-${color}`]: color,
      active,
      disabled,
    },
    disabled,
    ...rest,
  })
}
List.Item.displayName = 'ListItem'
