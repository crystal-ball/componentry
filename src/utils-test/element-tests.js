/* eslint-disable import/no-extraneous-dependencies */
import React from 'react'
import { mount } from 'enzyme'

import dt from './dt'

/**
 * Library default component requirment test suite.
 *
 * IMPORTANT NOTES
 * - Enzyme `find()` will find **EVERY** instance of a prop, this makes it difficult
 *   to assert on find lengths, because it will be variable for components with
 *   HOCs which is why we use `toBeTruthy` to only check that some instance of the
 *   expected element attr exists
 */

export default (TestComponent, enzymeOptions = {}) => {
  /*
   * Test that any html tag or component can be passed through the 'as' prop to
   * override the parent element the component renders
   */
  test('should render as specified html element or component', () => {
    // Create a component to validate that the TestComponent returns.
    const TestAs = () => <section />
    const wrapper = mount(<TestComponent as={TestAs} />, enzymeOptions)
    expect(wrapper.find(TestAs).length).toBeTruthy()
  })

  /*
   * All components should spread passed props on the parent container for max
   * customization
   */
  test('should pass through classnames and attributes', () => {
    const wrapper = mount(
      <TestComponent className="test-custom" data-test="test-custom" />,
      enzymeOptions,
    )
    expect(wrapper.find('.test-custom').length).toBeTruthy()
    expect(wrapper.find(dt('test-custom')).length).toBeTruthy()
  })

  // TODO: should render passed children
}
