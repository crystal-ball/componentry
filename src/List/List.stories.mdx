import { Meta, Story } from '@storybook/addon-docs'
import { BADGE } from '@geometricpanda/storybook-addon-badges'

import { colors } from '../storybook'
import { List } from './List'

<Meta
  title='Components/List'
  component={List}
  subcomponents={{ ListItem: List.Item }}
  parameters={{ badges: [BADGE.EXPERIMENTAL] }}
/>

export const Template = (args) => <List {...args} />

# ListGroup

## Unordered lists

<Story
  name='UnorderedList'
  args={{
    children: [
      <List.Item active>Radical React</List.Item>,
      <List.Item>Radical React</List.Item>,
      <List.Item>Component Library</List.Item>,
      <List.Item>Design System</List.Item>,
      <List.Item disabled>Radical React</List.Item>,
    ],
  }}
>
  {Template.bind({})}
</Story>

## Anchor list

<List>
  <List.Item href='#' active>
    Radical React
  </List.Item>
  <List.Item href='#'>Radical React</List.Item>
  <List.Item href='#'>Component Library</List.Item>
  <List.Item href='#'>Design System</List.Item>
  <List.Item href='#' disabled>
    Radical React
  </List.Item>
</List>

## Button list

<List>
  <List.Item active onClick={console.log}>
    Radical React
  </List.Item>
  <List.Item onClick={console.log}>Radical React</List.Item>
  <List.Item onClick={console.log}>Component Library</List.Item>
  <List.Item onClick={console.log}>Design System</List.Item>
  <List.Item disabled onClick={console.log}>
    Radical React
  </List.Item>
</List>

## Theme color items

<List>
  {colors.map((color) => (
    <List.Item key={color} color={color}>
      {color} themed list item
    </List.Item>
  ))}
</List>

## Flush lists

<List flush>
  <List.Item>God creates dinosaurs.</List.Item>
  <List.Item>God destroys dinosaurs.</List.Item>
  <List.Item>God creates Man.</List.Item>
  <List.Item>Man destroys God.</List.Item>
  <List.Item>Man creates Dinosaurs.</List.Item>
</List>
