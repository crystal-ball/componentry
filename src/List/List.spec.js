import React from 'react'
import { render } from '@testing-library/react'

import { elementTests } from '../../test/element-tests'
import { List } from './List'

describe('<List />', () => {
  // Basic library element test suite
  elementTests(List)
  elementTests(List.Item)
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<List /> snapshots', () => {
  test('renders ul list correctly', () => {
    const { container } = render(
      <List>
        <List.Item>Item 1</List.Item>
        <List.Item>Item 2</List.Item>
      </List>,
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  test('renders anchor list correctly', () => {
    const { container } = render(
      <List>
        <List.Item href='#'>Item 1</List.Item>
        <List.Item href='#'>Item 2</List.Item>
      </List>,
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  test('renders button list correctly', () => {
    const { container } = render(
      <List>
        <List.Item onClick={() => {}}>Item 1</List.Item>
        <List.Item onClick={() => {}}>Item 2</List.Item>
      </List>,
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
