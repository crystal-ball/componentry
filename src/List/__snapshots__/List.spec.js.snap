// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> snapshots renders correctly 1`] = `
<ul
  class="list-group"
>
  <li
    class="list-group-item"
  >
    Item 1
  </li>
  <li
    class="list-group-item"
  >
    Item 2
  </li>
</ul>
`;
